plutus-core/plutus-core/src/PlutusCore/Builtin/KnownType.hs

```diff
@@ -97,22 +97,46 @@ builtins this is rarely the case as most of the time we want aggressive inlining
 and the "just compute the damn thing" behavior.
 -}
 
-{- Note [Unlifting values of built-in types]
-It's trivial to unlift from a term a value of a monomorphic type like 'Integer': just check that
-the term is a constant, extract the type tag and check it for equality with the type tag of
-'Integer'.
+{- Note [Unlifting terminology]
+This function:
+
+    f :: Integer -> CkValue DefaultUni fun
+    f = VCon . Some . ValueOf DefaultUniInteger
+
+lifts an 'Integer' to 'CkValue'. Unlifting is the opposite:
+
+    g :: CkValue DefaultUni fun -> Maybe Integer
+    g (VCon (Some (ValueOf uni x))) = case uni of
+        DefaultUniInteger -> Just x
+        _                 -> Nothing
+
+The following usages of the "unlift" term are grammatical:
+
+1. unlift a 'CkValue' to 'Integer'
+2. unlift to 'Integer'
+3. unlift a 'CkValue' as an 'Integer'
+4. unlift from the 'VCon' constructor (or just 'VCon') to 'Integer'
+
+We call the integer that @g@ returns "the unlifted integer".
+-}
+
+{- Note [Unlifting a term as a value of a built-in type]
+See Note [Unlifting terminology] first.
+
+It's trivial to unlift a term to a monomorphic type like 'Integer': just check that the term is a
+constant, extract the type tag and check it for equality with the type tag of 'Integer'.
 
 Things work the same way for a fully monomorphized polymorphic type, i.e. @(Integer, Bool@) is not
 any different from just 'Integer' unlifting-wise.
 
-However there's no sensible way of unlifting a value of, say, @[a]@ where @a@ in not a built-in
-type. So let's say we instantiated @a@ to an @Opaque val rep@ like we do for polymorphic functions
-that don't deal with polymorphic built-in types (e.g. @id@, @ifThenElse@ etc). That would mean we'd
-need to write a function from a @[a]@ for some arbitrary built-in @a@ to @[Opaque val a]@. Which
-is really easy to do: it's just @map makeKnown@. But the problem is, unlifting is supposed to be
-cheap and that @map@ is O(n), so for example 'MkCons' would become an O(n) operation making
-perfectly linear algorithms quadratic. See https://github.com/input-output-hk/plutus/pull/4215 for
-how that would look like.
+However there's no sensible way of unlifting to, say, @[a]@ where @a@ in not a built-in type. So
```

Comment:
Can we give a more precise description (or point to it if it already exists somewhere) of what can and cannot be unlifted to, rather than just giving one positive example (`Integer`) and one negative example (`[a]`). E.g., you can't unlift in the following cases, I think:

- Types that contain any non-builtin type or tycon (e.g., `Maybe whatever`), except `Opaque`, `SomeConstant`, `EvaluationResult`
- `Opaque`, `SomeConstant`, `EvaluationResult` appearing within a built-in tycon, e.g., `[SomeConstant uni [Integer]]` (technically this can be unlifted but is inefficient)
- ?